Compiler/src/profiling/tracy.jl

```diff
@@ -0,0 +1,67 @@
+module Tracy
+
+import ..@noinline, ..Cint, ..Vector, ..push!, ..unsafe_convert, ..esc,
+       ..pointer_from_objref, ..String, ..Ptr, ..UInt8, ..Cvoid,
+       ..Expr, ..LineNumberNode, ..Symbol, ..UInt32, ..C_NULL, ..(===)
+
+_strpointer(s::String) = ccall(:jl_string_ptr, Ptr{UInt8}, (Any,), s)
+
+mutable struct TracySrcLoc
+    zone_name::Ptr{UInt8}
+    function_name::Ptr{UInt8}
+    file::Ptr{UInt8}
+    line::UInt32
+    color::UInt32
+    # Roots
+    zone_name_str::String
+    function_name_sym::Symbol
+    file_sym::Symbol
+end
+TracySrcLoc(zone_name::String, function_name::Symbol, file::Symbol, line::UInt32, color::UInt32) =
+    TracySrcLoc(C_NULL, C_NULL, C_NULL, line, color, zone_name, function_name, file)
+
+@noinline function reinit!(srcloc::TracySrcLoc)
+    lock_timing()
```

Comment:
I don't think you need the lock here - it's OK for the threads to race since the outcome is idempotent.

It's also insufficient to prevent UB, since you have an unguarded read in begin_event. I'd recommend `@atomic` set/get on these fields instead, which should fix both issues